snapshots.go

```diff
@@ -621,3 +622,262 @@ func (s *Stream) SnapshotToDirectory(ctx context.Context, dir string, opts ...Sn
+	defer trackingR.Close()
+	defer trackingW.Close()
+	go progress.trackBlockProgress(trackingR, sopts.debug, errc)
+
+	writer := io.MultiWriter(dataBuffer, trackingW)
+
+	// tell the caller we are starting and what to expect
+	progress.notify()
+
+	sub, err := s.mgr.nc.Subscribe(ib, func(m *nats.Msg) {
+		if len(m.Data) == 0 {
+			m.Sub.Unsubscribe()
+			cancel()
+			return
+		}
+
+		progress.Lock()
+		progress.bytesReceived += uint64(len(m.Data))
+		progress.chunksReceived++
+		progress.Unlock()
+
+		n, err := writer.Write(m.Data)
+		if err != nil {
+			errc <- err
+			return
+		}
+		if n != len(m.Data) {
+			errc <- fmt.Errorf("failed to write %d bytes to %s, only wrote %d", len(m.Data), sopts.dataFile, n)
+			return
+		}
+
+		if m.Reply != "" {
+			if sopts.debug {
+				progress.Lock()
+				log.Printf("Responding to server subject %s %s chunks, last chunk size %s", m.Reply, humanize.Comma(int64(progress.chunksReceived)), humanize.IBytes(uint64(len(m.Data))))
+				progress.Unlock()
+			}
+
+			m.Respond(nil)
+		}
+	})
+	if err != nil {
+		return progress, err
+	}
+	defer sub.Unsubscribe()
+	sub.SetPendingLimits(-1, -1)
+
+	select {
+	case err := <-errc:
+		if sopts.debug {
+			log.Printf("Snapshot Error: %s", err)
+		}
+
+		return progress, err
+	case <-sctx.Done():
+		meta := map[string]any{
+			"config": resp.Config,
+			"state":  resp.State,
+		}
+		mj, err := json.MarshalIndent(meta, "", "  ")
+		if err != nil {
+			return nil, err
+		}
+
+		metadataBuffer.Write(mj)
+
+		progress.finished = true
+		progress.notify()
+
+		return progress, nil
+	}
+}
+
+func (m *Manager) RestoreSnapshotFromBuffer(ctx context.Context, stream string, dataReader, metadataReader io.Reader, opts ...SnapshotOption) (RestoreProgress, *api.StreamState, error) {
+	sopts := &snapshotOptions{
+		chunkSz: 64 * 1024,
+	}
+
+	for _, opt := range opts {
+		opt(sopts)
+	}
+
+	req := api.JSApiStreamRestoreRequest{}
+	mj, err := io.ReadAll(metadataReader)
+	if err != nil {
+		return nil, nil, err
+	}
+	err = json.Unmarshal(mj, &req)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// allow full config override
+	if sopts.restoreConfig != nil {
+		req.Config = *sopts.restoreConfig
+	}
+
+	// allow just stream name override
+	//
+	// this used to be allowed but turns out the server support for this was
+	// not up to scratch and fixing it would mean having to rebuild and re-checksum
+	// every message, so for now we error here instead
+	if req.Config.Name != stream {
+		return nil, nil, fmt.Errorf("stream name may not be changed during restore")
+	}
+
+	if req.Config.Storage == api.MemoryStorage {
+		return nil, nil, ErrMemoryStreamNotSupported
+	}
+
+	var resp api.JSApiStreamRestoreResponse
+	err = m.jsonRequest(fmt.Sprintf(api.JSApiStreamRestoreT, req.Config.Name), req, &resp)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	progress := &snapshotProgress{
+		startTime: time.Now(),
+		chunkSize: sopts.chunkSz,
+		// chunksToSend: 1 + int(fstat.Size())/sopts.chunkSz,
```

Comment:
I wasn't sure how to resolve this since we don't have the total size of the snapshot anymore given we're reading the snapshot in from an `io.Reader`

Reply:
Could we accept the size as an argument in the buffered version of this? I imagine you'd end up running this unattentended anyway so maybe having progress is not useful in that case and you could skip the entire progress handling?